omegaconf/dictconfig.py

```diff
@@ -574,6 +574,7 @@ def _set_value(self, value: Any, flags: Optional[Dict[str, bool]] = None) -> Non
             self.__dict__["_content"] = "???"
             self._metadata.object_type = None
         else:
+            previous_state = copy.copy(self.__dict__["_content"])
```

Comment:
Why do you need to copy it? it's expensive and I am not seeing the value in that.

Reply:
Content is changed to {}, that means if it isn't copied it will still be modified.

Reply:
copying the entire subtree on assignment is expensive.
content is ASSIGNED {}, if you keep a reference to the previous one is it changing? is this something you have actually seen or are you just thinking it will change?

Reply:
ah, I got it.
I think this is probably okay because it's a shallow copy, but please do some benchmarking of the initialization of OmegaConf from a fairly big dictionary to ensure this does not introduce a regression speed. we should over time improve performance or at least not make it slower.

Reply:
With a dict of 1M key-value pair.

```python
@dataclass
class DictClass:
    foo: Dict[str, int] = field(default_factory=lambda: {"a": 4})

cols = [str(i) for i in range(1000000)]
data = [i for i in range(1000000)]
dict_ = dict(zip(cols, data))
cfg = OmegaConf.structured(DictClass)

t1 = datetime.now()
cfg.foo = dict_
t2 = datetime.now()
cfg.foo = {"a": 1}
t3 = datetime.now()

print("Execution time: ", (t2-t1))
print("Assigning new value(it'll shallow copy content) time: ", t3-t2)
```
**Removing the shallow copy**
<img width="892" alt="image" src="https://user-images.githubusercontent.com/30913090/97086154-eebc9400-1621-11eb-9380-7ac368e90252.png">

**With the shallow copy**
<img width="892" alt="image" src="https://user-images.githubusercontent.com/30913090/97086150-e5332c00-1621-11eb-895e-fa95c0e90182.png">
As you can see there isn't a significative difference between both implementations.

Reply:
Check a dictionary with some nesting.
How about 10 levels, each one with two splits (should be about ~1000 leaf nodes).

```yaml
tree:
  left:
    left: 1
    right: 1
  right:
    left: 1
    right: 1
```

Reply:
With and without it causes segmentation fault.

Reply:
also, take an average of many iterations.

Reply:
1024 leaf nodes
**Without copy**
<img width="594" alt="image" src="https://user-images.githubusercontent.com/30913090/97112003-3d316780-16e2-11eb-979b-1d3cf592f89b.png">
**With copy**
<img width="594" alt="image" src="https://user-images.githubusercontent.com/30913090/97112170-19baec80-16e3-11eb-9a25-c5e4c72325ad.png">
It looks like it doesn't change anything. I can do further statistics if you prefer.

Reply:
No, it sounds good. can you paste your script here or in a GitHub gist?

Reply:
https://gist.github.com/pereman2/1b9069bf68a2df2a8023dcf34fd4147d

Reply:
Your benchmark is checking assignment, my concern is about initialization of configs like that one.

Reply:
I'll test creating.